### DISCUSSION TITLE
read in a2l file

### DISCUSSION THREAD

**Discussion by Mrc1979** [christoph2/pyA2L#31] [Q&A]
Hi Christoph,

I wanted to execute your example of the doc.
Unfortunately directly an error occurs when I try to read the file

from pya2l import DB

db = DB()
session = db.import_a2l(r"dat/vehicleEcu.a2l")

Is there any package or module missing or why that class is not available?

Thanks in advance for your answer.

Greetings Maurice

**Comment by christoph2** [comment]
Could you post the exception trace?

**Reply by Mrc1979** [reply]
Hi Christoph,

here it is - the DB-classseems to be missing:

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_21640/433800489.py in <module>
----> 1 from pya2l import DB
      2 db = DB()
      3 
      4 session = db.import_a2l(r"dat/vehicleEcu.a2l")
      5 

ImportError: cannot import name 'DB' from 'pya2l' (C:\Users\blm5si\.conda\envs\a2l\lib\site-packages\pya2l\__init__.py)

**Comment by christoph2** [comment]
Hmm, that's a communication problem of mine...
After things got more serious, I noticed that the name pya2l was already taken...
The correct pip name is now
      pya2ldb
I'll update the documentation accordingly, and yes, the package name should also be fixed, but that's a code breaking change.

**Reply by Mrc1979** [reply]
Hey Christoph, sounds great if your implementation is available under that name - I directly will try :).
Thanks for your response! Greetings Maurice

**Reply by Mrc1979** [reply]
Hi Christoph. Thanks for the package name hint. I installed now correctly and it' callable.
But I get that error:...

Have you any idea how to solve / work around?

Thanks Maurice

PS.: I run it in Win10x64 - or should I under Linux?



---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_13668/433800489.py in <module>
      2 db = DB()
      3 
----> 4 session = db.import_a2l(r"dat/myEcuFile.a2l")
      5 

~\.conda\envs\a2l\lib\site-packages\pya2l\__init__.py in import_a2l(self, file_name, debug, in_memory, remove_existing, encoding)
    113                 raise OSError("file '{}' already exists.".format(self._dbfn))
    114         prepro = Preprocessor()
--> 115         prepro_result = prepro.process(self._a2lfn, encoding = detect_encoding(self._a2lfn))
    116         a2l_parser = ParserWrapper("a2l", "a2lFile", A2LListener, debug = debug, line_map = prepro_result.line_map)
    117 

~\.conda\envs\a2l\lib\site-packages\pya2l\preprocessor.py in process(self, file_name, encoding)
    158         self.aml_section = None
    159         self.if_datas = []
--> 160         data = self._process_file(file_name, join_lines = True, encoding = encoding)
    161         return PreprocessorResult(self._process_aml(data), self.aml_section, self.if_data_sections, LineMap(self.line_map))
    162 

~\.conda\envs\a2l\lib\site-packages\pya2l\preprocessor.py in _process_file(self, file_name, join_lines, encoding)
    171             raise RuntimeError("Circular dependency to include file '{}'.".format(abs_file_name))
    172         self.logger.debug("Processing file '{}'.".format(abs_file_name))
--> 173         for num, line in enumerate(f_obj, 1):
    174             self.absolute_file_number += 1
    175             line = line.rstrip("\n")

~\.conda\envs\a2l\lib\encodings\ascii.py in decode(self, input, final)
     24 class IncrementalDecoder(codecs.IncrementalDecoder):
     25     def decode(self, input, final=False):
---> 26         return codecs.ascii_decode(input, self.errors)[0]
     27 
     28 class StreamWriter(Codec,codecs.StreamWriter):

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 6581: ordinal not in range(128)

**Comment by Mrc1979** [comment]
Ahhhhhh-  I checked the ASAP2-file.
Perhaps the xB0, xF6, xDF... causes troubles? That the German ö, ä, ß, ü...

https://www.codetable.net/hex/e4

Idea: We could in before parsing/processing operation via re.replace(...) exchange the literals.

[0xc4, 0xe4, 0xd6, 0xf6, 0xdc, 0xfc, 0xdf] also:
[0xC4, 0xE4, 0xD6, 0xF6, 0xDC, 0xFC, 0xDF]
Ä, ä, Ö, ö, Ü, ü, ß

**Reply by Mrc1979** [reply]
Solution:
----------------
import pathlib
from pya2l import DB

# Read file as string
a2lcontent_str = open(r'dat/myFile.a2l', 'r').read()

# Transform unicodes (latin1)
char_map = {ord('ä'):'ae', ord('ü'):'ue', ord('ö'):'oe', ord('ß'):'ss', ord('µ'):'mu' \
           , ord('Ä'):'Ae', ord('Ö'):'Oe', ord('Ü'):'Ue'}

# Save as new file w.o. unwanted unicodes
a2lcontentCleaned_str = a2lcontent_str.translate(char_map)
pathlib.Path("dat/output.a2l").write_text(a2lcontentCleaned_str)

# run parser
db = DB()
session = db.import_a2l(r"dat/output.a2l")

**Reply by Mrc1979** [reply]
Better: Internal parsing, catching error when line-parsing and replacing internally. ;)

**Comment by christoph2** [comment]
No -- Unicode support is required by the standard, we can't just cut out the offending chars...
Hmm, I'm using A2Ls containing German umlauts, but I've never had any problems, strange.  The latest version has fixed the encoding of the parser, but the issue seams to be located in the pre-processor, I'll do some further investigations.
Btw: You may check your /include statements, the pre-processor reports a circular dependency.
P.S.: Yes, A2L runs on Windows and Linux (e.g RaspPI).

**Reply by karsten-burger** [reply]
Hello, I just wanted to use your library but ran straight into the problem with Unicode characters. Is there any simple solution?
